### PR TITLE
#166383226 Build Identification helper

### PR DESCRIPTION
**Pull request by henryade**
#### What does this PR do?
Adds identification middleware

#### Description of Task to be completed?
Build identification middelware to identify users by id or email

#### How should this be manually tested?
N/A

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#166383226](https://www.pivotaltracker.com/story/show/166383226)

#### Screenshots (if appropriate)
N/A

#### Questions:
N/A

### CODE REVIEW

File: server/database/models/author.js [hunk 1]
```diff
@@ -1,6 +1,5 @@
 export default (sequelize, DataTypes) => {
   const Author = sequelize.define('Author', {
-    id: DataTypes.INTEGER,
```
Comment (KvNGCzA) [inline server/database/models/author.js:3]: Please do not take out the ids.

File: server/helpers/identification.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { UserRole, User, Role } from '../database/models';
+
+const findIdByEmail = email => User.findOne({
```
Comment (KvNGCzA) [inline server/helpers/identification.js]: This function should take in response as an argument and use it to return proper error messages.

**Review by KvNGCzA — CHANGES_REQUESTED** [review]

File: server/database/models/author.js (hunk 1 above)
Comment (henryade) [reply server/database/models/author.js:3]: > Please do not take out the ids.

Okay. Done

File: server/helpers/identification.js (hunk 2 above)
Comment (henryade) [reply server/helpers/identification.js]: > This function should take in response as an argument and use it to return proper error messages.

Done

File: server/helpers/identification.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import { UserRole, User, Role } from '../database/models';
+
+const findIdByEmail = (email, response) => User.findOne({
+  where: {
+    email,
+  },
+  attributes: ['id']
```
Comment (KvNGCzA) [inline server/helpers/identification.js]: Why are you returning just the id?

File: server/helpers/identification.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+})
+  .then((user) => {
```
Comment (KvNGCzA) [inline server/helpers/identification.js]: use an async function, not then and catch.

File: server/helpers/identification.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    if (!user) {
+      return response.status(404).json({
```
Comment (KvNGCzA) [inline server/helpers/identification.js]: use the responseMessage helper function

File: server/helpers/identification.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import { UserRole, User, Role } from '../database/models';
+
+const findIdByEmail = (email, response) => User.findOne({
```
Comment (KvNGCzA) [inline server/helpers/identification.js]: Rename this function better

File: server/helpers/identification.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import { UserRole, User, Role } from '../database/models';
```
Comment (KvNGCzA) [inline server/helpers/identification.js]: Have you tested this function?

File: server/helpers/identification.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      return response.status(404).json({
+        status: 'Failure',
+        error: 'User not found',
+      });
+    }
+    return user.dataValues.id;
+  })
+  .catch(error => error);
+
+export default async (param, response) => {
```
Comment (KvNGCzA) [inline server/helpers/identification.js]: rename this function better

**Review by KvNGCzA — CHANGES_REQUESTED** [review]
Use only destructuring through out the codebase

File: server/helpers/identification.js (hunk 3 above)
Comment (henryade) [reply server/helpers/identification.js]: > Why are you returning just the id?

I am using the email to get the id of the account

**Review by Adekoreday — CHANGES_REQUESTED** [review]
good work

**Review by Adekoreday — CHANGES_REQUESTED** [review]
in your helper/identification.js
i think you can make your 
include: [
{model: Role, required: true }
]

so it can return user if only its has a role

**Comment by henryade** [comment]
> in your helper/identification.js
> i think you can make your
> include: [
> {model: Role, required: true }
> ]
> 
> so it can return user if only its has a role

Done

File: server/helpers/identification.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import { UserRole, User, Role } from '../database/models';
+import responseMessage from './responseMessage';
+
+const getUserId = async (email, response) => {
+  try {
+    const { dataValues } = await User.findOne({ where: { email } });
+    return dataValues.id;
+  } catch (error) {
+    responseMessage(response, 404, { message: 'Account not found', error });
```
Comment (Ukhu) [inline server/helpers/identification.js]: I don't think you need to return the error object to the user. The message already tells the user what has happened.

Comment (henryade) [reply server/helpers/identification.js]: Done

**Review by Adekoreday — CHANGES_REQUESTED** [review]
nice work

**Comment by henryade** [comment]
ok my bad

It's getting it from the userrole table not the role table.

**Review by Adekoreday — DISMISSED** [review]

File: server/helpers/responseMessage.js [hunk 10]
```diff
@@ -1,6 +1,6 @@
-export default (response, statusCode, object) => {
+export default (response, statusCode, object, status = 'failure') => {
```
Comment (Ukhu) [inline server/helpers/responseMessage.js]: Don't modify this function. You can add the new status to the object you are passing to the function, rather than passing the new status as a separate argument

File: server/helpers/responseMessage.js [hunk 11]
```diff
@@ -1,6 +1,6 @@
...
   response.status(statusCode).json({
-    status: 'failure',
```
Comment (Ukhu) [inline server/helpers/responseMessage.js]: Revert the changes in this function. It shouldn't even show up in your changed files.

File: server/helpers/identification.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  }
+};
+const getUserRole = async (param, response) => {
+  try {
+    const userId = (/\D/g.test(param)) ? await getUserId(param, response) : param;
+    const { dataValues } = await UserRole.findOne({
+      where: { userId },
+      include: [{ model: Role, required: true }]
+    });
+    return responseMessage(response, 200, { data: dataValues.Role.dataValues }, 'success');
```
Comment (Ukhu) [inline server/helpers/identification.js]: Add the status to the object. Don't pass it as a fourth argument.

File: server/helpers/responseMessage.js (hunk 10 above)
Comment (henryade) [reply server/helpers/responseMessage.js]: If I add it, it will show the failure status first.

File: server/helpers/responseMessage.js (hunk 11 above)
Comment (henryade) [reply server/helpers/responseMessage.js]: I was asked to use the responseMessage function. The only way to show success is either to create a new function which will lead to duplication or to add that.

File: server/helpers/identification.js (hunk 12 above)
Comment (henryade) [reply server/helpers/identification.js]: If I add it to the object, it will show the failure status first

Comment (Ukhu) [reply server/helpers/identification.js]: No, it will override the failure status already there. Just put it as the first property in the object

File: server/helpers/responseMessage.js (hunk 10 above)
Comment (Ukhu) [reply server/helpers/responseMessage.js]: Just put it as the first property in the object. It will override the already existing status.

File: server/helpers/responseMessage.js (hunk 11 above)
Comment (Ukhu) [reply server/helpers/responseMessage.js]: Please check my latest comment above

File: server/helpers/identification.js (hunk 12 above)
Comment (henryade) [reply server/helpers/identification.js]: Done

File: server/helpers/responseMessage.js (hunk 11 above)
Comment (henryade) [reply server/helpers/responseMessage.js]: Done

File: server/helpers/responseMessage.js (hunk 10 above)
Comment (henryade) [reply server/helpers/responseMessage.js]: Done

**Review by JuwonAbiola — DISMISSED** [review]
good

File: server/helpers/identification.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import { User } from '../database/models';
+import responseMessage from './responseMessage';
+
+const getUserId = async (param, response) => {
+  try {
+    const field = (/\D/g.test(param)) ? { email: param } : { id: param };
```
Comment (KvNGCzA) [inline server/helpers/identification.js:8]: what happens when there is a number in the email?

File: server/helpers/identification.js [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import { User } from '../database/models';
+import responseMessage from './responseMessage';
+
+const getUserId = async (param, response) => {
```
Comment (KvNGCzA) [inline server/helpers/identification.js]: rename this function, and ensure to name the file after its default export

File: server/helpers/identification.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+  try {
+    const field = (/\D/g.test(param)) ? { email: param } : { id: param };
+    const user = await User.findOne({ where: field });
```
Comment (KvNGCzA) [inline server/helpers/identification.js:9]: What happens when this doesn't find a user?

File: server/helpers/identification.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
+import { User } from '../database/models';
```
Comment (KvNGCzA) [inline server/helpers/identification.js]: import `models`, then on a new line destructure `const { User } = models;`

File: server/helpers/index.js [hunk 17]
```diff
@@ -1,7 +1,9 @@
 import createToken from './createToken';
 import responseMessage from './responseMessage';
+import identify from './identification';
 
 export default {
   createToken,
-  responseMessage
+  responseMessage,
+  identify
```
Comment (KvNGCzA) [inline server/helpers/index.js]: Find a better and more descriptive name than `identify`

**Review by KvNGCzA — CHANGES_REQUESTED** [review]

File: server/helpers/identification.js (hunk 13 above)
Comment (henryade) [reply server/helpers/identification.js:8]: it will still see it as an email

File: server/helpers/identification.js (hunk 16 above)
Comment (henryade) [reply server/helpers/identification.js]: Done

File: server/helpers/identification.js (hunk 15 above)
Comment (henryade) [reply server/helpers/identification.js:9]: it returns undefined then whoever wants to use the function will check for undefined.

File: server/helpers/identification.js (hunk 14 above)
Comment (henryade) [reply server/helpers/identification.js]: Done

File: server/helpers/index.js (hunk 17 above)
Comment (henryade) [reply server/helpers/index.js]: changed

File: server/helpers/identification.js (hunk 13 above)
Comment (henryade) [reply server/helpers/identification.js:8]: the only time it regular expression returns false is if the param is a number

**Review by Ukhu — APPROVED** [review]

**Review by KvNGCzA — APPROVED** [review]
